### PR TITLE
Multithread support for wolfMQTT-SN

### PR DESCRIPTION
**Pull request by embhorn**
Adding multithread support to the sensor network client library. Also adds a new SN multithread example that demonstrates the API. Configure with `--enable-sn --enable-mt`. Tested with QoS 0,1, and 2.

Also fixes an issue with subscribe using short names.

### CODE REVIEW

File: src/mqtt_client.c [hunk 1]
```diff
@@ -2450,6 +2548,16 @@ static int SN_Client_WaitType(MqttClient *client, void* packet_obj,
...
     packet_type = SN_MSG_TYPE_RESERVED;
+#ifdef WOLFMQTT_MULTITHREAD
+    pendResp = NULL;
+    readLocked = 0;
+#endif
+    waitMatchFound = 0;
+
+#ifdef WOLFMQTT_DEBUG_CLIENT
+    PRINTF("SN_Client_WaitType: Type %s (%d), ID %d",
+            SN_Packet_TypeDesc(wait_type), wait_type, wait_packet_id);
```
Comment (dgarske) [inline src/mqtt_client.c:2559]: If building with `CC=g++` get a cast issue here. Should be `SN_Packet_TypeDesc((SN_MsgType)wait_type)`.

Same for line 736 case above. Needs to be `MqttPacket_TypeDesc((MqttPacketType)wait_type)`.

File: src/mqtt_client.c [hunk 2]
```diff
@@ -2974,19 +3801,76 @@ int SN_Client_Disconnect_ex(MqttClient *client, SN_Disconnect *disconnect)
...
+        }
+    #endif
+        return rc;
+    }
 
     /* If sleep was set, wait for response disconnect packet */
     if ((disconnect != NULL) && (disconnect->sleepTmr != 0)) {
         rc = SN_Client_WaitType(client, NULL,
                 SN_MSG_TYPE_DISCONNECT, 0, client->cmd_timeout_ms);
+    #ifdef WOLFMQTT_NONBLOCK
```
Comment (dgarske) [inline src/mqtt_client.c:3864]:
```
src/mqtt_client.c:3863:9: warning: Value stored to 'rc' is never read
        rc = SN_Client_WaitType(client, NULL,
        ^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/sn-client/sn-client.c [hunk 3]
```diff
@@ -289,9 +289,10 @@ int sn_test(MQTTCtx *mqttCtx)
     }
 
     {
-        /* Short Topic Name Subscribe */
+        /* Short topic name subscribe */
```
Comment (dgarske) [inline examples/sn-client/sn-client.c:292]: Above subscribe warning:

```
examples/sn-client/sn-client.c:254:9: warning: Value stored to 'rc' is never read
        rc = SN_Client_Subscribe(&mqttCtx->client, &subscribe);
        ^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by dgarske — CHANGES_REQUESTED** [review]
All code looks good and functionally tests out against SN-Gateway/Broker. Just a couple minor warnings I found (some old). Thanks!

File: examples/sn-client/sn-client.c (hunk 3 above)
Comment (embhorn) [reply examples/sn-client/sn-client.c:292]: Couldn't reproduce this one with `CC=g++`, but I fixed it anyway.

**Review by dgarske — APPROVED** [review]

**Pull request merged**